lib/rubygems/version.rb

```diff
@@ -423,16 +412,12 @@ def _segments
     # segments is lazy so it can pick up version values that come from
     # old marshaled versions, which don't go through marshal_load.
     # since this version object is cached in @@all, its @segments should be frozen
+    @segments ||= partition_segments(@version)
+  end
 
-    @segments ||= @version.scan(/[0-9]+|[a-z]+/i).map! do |s|
+  def partition_segments(ver)
+    ver.scan(/\d+|[a-z]+/i).map! do |s|
       /^\d+$/.match?(s) ? s.to_i : -s
```

Comment:
```suggestion
      /\A\d/.match?(s) ? s.to_i : -s
```

Reply:
This is a little slower than /\d/ which still achieves the same purpose. (it can't be anything other than a string of all digits or a string of all alphas).

Reply:
hm, I have matching an anchored `\d` being faster (which makes sense, `/\d/` is O(n) to fail matching, whereas the anchored version is constant time)

Reply:
weird, my synthetic test must not be as good as yours. Let's go with the anchored one.